File: lint-arwaky/src-python/taxonomy/semantic_tracer_error.py
"""semantic_tracer_error — Semantic analysis domain error types."""

from pydantic import BaseModel, ConfigDict

from .error_code_vo import ErrorCode
from .error_value_vo import (
    ErrorMessage,
    Cause,
)
from .file_path_vo import FilePath


class SemanticError(BaseModel):
    """General failure during semantic code analysis."""

    model_config = ConfigDict(frozen=True)

    path: FilePath | None = None
    message: ErrorMessage
    error_code: ErrorCode | None = None
    cause: Cause | None = None

    def __str__(self):
        target = f" on {self.path}" if self.path else ""
        code = f" [{self.error_code}]" if self.error_code else ""
        return f"Semantic Error{target}{code}: {self.message}"


class ScopeResolutionError(SemanticError):
    """Failed to resolve or identify code scope (class/function)."""
    pass


class CallChainError(SemanticError):
    """Failed to trace or build a call graph/chain."""
    pass

File: lint-arwaky/src-rust/taxonomy/semantic_tracer_error.rs
use serde::{Serialize, Deserialize};
use super::*;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct SemanticError {
    #[serde(default)]
    pub path: Option<FilePath>,
    pub message: ErrorMessage,
    #[serde(default)]
    pub error_code: Option<ErrorCode>,
    #[serde(default)]
    pub cause: Option<Cause>,
}

impl SemanticError {
    pub fn new(message: ErrorMessage) -> Self {
        Self { path: None, message, error_code: None, cause: None }
    }
}

impl std::fmt::Display for SemanticError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let target = self.path.as_ref().map(|p| format!(" on {}", p)).unwrap_or_default();
        let code = self.error_code.as_ref().map(|c| format!(" [{}]", c)).unwrap_or_default();
        write!(f, "Semantic Error{}{}: {}", target, code, self.message)
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ScopeResolutionError {
    #[serde(flatten)]
    pub base: SemanticError,
}

impl ScopeResolutionError {
    pub fn new(message: ErrorMessage) -> Self {
        Self { base: SemanticError::new(message) }
    }
}

impl std::fmt::Display for ScopeResolutionError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.base)
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct CallChainError {
    #[serde(flatten)]
    pub base: SemanticError,
}

impl CallChainError {
    pub fn new(message: ErrorMessage) -> Self {
        Self { base: SemanticError::new(message) }
    }
}

impl std::fmt::Display for CallChainError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.base)
    }
}
